ERCS/erc-7913.md

```diff
@@ -0,0 +1,112 @@
+   * SHOULD return 0xffffffff or revert if the key is empty
+   */
+  function verify(bytes calldata key, bytes32 hash, bytes calldata signature) external view returns (bytes4);
+}
+```
+
+Verifiers SHOULD be stateless.
+
+## Rationale
+
+Verifiers can be used to avoid deploying many ERC-1271 "identity contract" (one per key), which would be expensive. Using this model, new keys can be used without any deployment costs.
+
+ERC-1271 already covers the cases were a key is not necessary. To avoid ambiguity, the verifiers contracts should not support (or be expected to support) empty keys. Signers that are 20 bytes long (empty key) should be handled using ERC-1271.
+
+Consistency with existing systems (ecrecover and ERC-1271) requires the message to be `bytes32` hash. These are usually produced following [ERC-191](./eip-191.md) or [ERC-712](./eip-712.md). Cryptographic systems that use different hashing methods SHOULD see this hash as a message, and possibly rehash it following the relevant standards.
+
+## Backwards Compatibility
+
+A system can support [ERC-7913](./eip-7913.md) signers alongside EOAs and ERC-1271 in the following way.
+
+A signer is a `bytes` object that is the concatenation of an address and optionally a key: `verifier || key`. A signer is at least 20 bytes long.
+
+Given a signer `signer`, a message hash `hash`, and a signature `sign`, verification is done as follows:
+
+- if `signer.length < 20`: verification fails;
+- split `signer` into `(verifier, key)` with `verifier` being the first 20 bytes and `key` being the rest (potentially empty)
+- if `key` is empty, then consider that `verifier` is the identity.
+  - verification is done using ERC-1271's isValidSignature if there is code at `verifier` address, or ecrecover otherwise
+- if `key` is not empty, call `IERC7913SignatureVerifier(verifier).verify(key, hash, signature)`
+  - if the return value is the expected magic value () then verification is successful,
+  - otherwise, verification fails.
+
+## Reference Implementation
+
+In solidity, signature verification COULD be implemented in the following library:
+
+```solidity
+import {SignatureChecker} from '@openzeppelin/contracts/utils/cryptography/SignatureChecker.sol';
+
+/// @dev Extention of openzeppelin's SignatureChecker library
+library SignatureCheckerExtended {
+  function isValidSignatureNow(bytes calldata signer, bytes32 hash, bytes memory signature) internal view returns (bool) {
+      if (signer.length < 20 ) {
+        return false;
+      } else if (signer.length == 20) {
+        return SignatureChecker.isValidSignatureNow(address(bytes20(signer)), hash, signature);
+      } else {
+        try IERC7913SignatureVerifier(address(bytes20(signer[0:20]))).verify(signer[20:], hash, signature) returns (bytes4 magic) {
+          return magic == IERC7913SignatureVerifier.verify.selector;
+        } catch {
+          return false;
+        }
+      }
+  }
+}
+```
+
+## Security Considerations
+
+TODO
```

Comment:
Author made a explict choice that "Verifiers SHOULD be stateless.", in which case `verify` will be `view/pure` and not causing state change. I'd suggest we remind compliant contract developer to add logic to prevent replay attack: such as if their application rely on the state such as fund availablity, nonce, or whether another internal TX has been performed.